[DOC-124] Admin API (#56)
Ed-Fi-Alliance-OSS/ed-fi-alliance-oss.github.io: docs/reference/3-admin-api/installation/admin-api-1x-for-odsapi-5x-6x/admin-api-1x-docker-installation.md

# Admin API 1.x - Docker installation

**Contents:**

* [Before You Install](#before-you-install)
  * [Compatibility & Supported ODS / API Versions](#compatibility-supported-ods-api-versions)
* [Installation Instructions](#installation-instructions)
  * [General Prerequisites](#general-prerequisites)
* [Installation Instructions](#installation-instructions)
  * [1\. Include Admin API in the ODS Docker Setup](#1-include-admin-api-in-the-ods-docker-setup)
  * [2\. Relaunch the Docker Composition](#2-relaunch-the-docker-composition)

>   * [3\. Execute First-Time Configuration](#3-execute-first-time-configuration)

Reviewer: In the ODS/API documentation, I have been removing these introductory TOC's, since Docusaurus builds a TOC sidebar on the right automatically. That's up to you though.

# Before You Install

This section provides general information you should review before installing the Ed-Fi ODS / API Admin API for v1.4.0.

## Compatibility & Supported ODS / API Versions

This version of the Admin API has been tested and can be installed for use with the Ed-Fi ODS / API v3.4 - 6.1. See the [Ed-Fi Technology Version Index](https://edfi.atlassian.net/wiki/spaces/ETKB/pages/20875717/Ed-Fi+Technology+Version+Index) for more details.

# Installation Instructions

## General Prerequisites

The following are required to install the Admin API:

> * The Admin API provides an interface to administer an Ed-Fi ODS / API. Understandably, you must have an instance of the Ed-Fi ODS / API v3.4 - 6.1 deployed and operational before you can use the Admin API. Tested configurations include on-premises installation via [binary installation](https://edfi.atlassian.net/wiki/spaces/ODSAPIS3V520/pages/25100419/Getting+Started+-+Binary+Installation) or [source code installation](https://edfi.atlassian.net/wiki/spaces/ODSAPIS3V520/pages/25100348/Getting+Started+-+Source+Code+Installation).

Reviewer: These pages have not been migrated yet. In a few weeks I will do a search for all Confluence links and try to fix any that should be changed.

* A SQL Server 2012 or higher, or Postgres 11 or higher database server (i.e., the same platform requirement applicable to your ODS / API).
* A modern web browser such as Google Chrome, Mozilla Firefox, or Microsoft Edge is required to view live Swagger documentation. Internet Explorer 11 (a pre-installed browser on Windows Server) may load but may not function when using Admin API.

# Installation Instructions

Admin API is not included with the ODS-Docker solution by default, but can be hosted as part of that ecosystem.

> To install Admin API on Docker, first Install the [ODS / API Docker](https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS-Docker) environment [following these instructions](https://edfi.atlassian.net/wiki/display/EDFITOOLS/Docker+Deployment). Then, apply the below changes to the environment to introduce the Admin API.  Admin API does not support in-place upgrades from prior versions.  Please install a fresh copy of Admin API to upgrade from prior versions.

Reviewer: ```suggestion
To install Admin API on Docker, first Install the [ODS / API Docker](https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS-Docker) environment [following these instructions](/reference/docker). Then, apply the below changes to the environment to introduce the Admin API.  Admin API does not support in-place upgrades from prior versions.  Please install a fresh copy of Admin API to upgrade from prior versions.
```

The new link is just a placeholder, but might as well get the change in there now.

> ## 1\. Include Admin API in the ODS Docker Setup

Reviewer: For some reason the Confluence export often added backslashes in front of periods. I _think_ this will be harmless, but it is worth double-checking that when you run the application.

### Docker Compose

Add the following to your `docker-compose.yml`  file. This can be done either instead of or in addition to the `adminapp`  service.

#### Admin API Application

This service depends on the `pb-admin`  and subsequently `db-admin` services to run.

> **docker-compose.yml**
> 
> ```

Reviewer: ```suggestion
```yaml title="docker-compose.yml"
```

This will give YML syntax highlighting, and the code block will have `docker-compose.yml` as a header/title.

**If you are introducing Admin API to an existing composition do** NOT **change the volume mapping configuration in order to preserve your data**.**** Only change the image and tag of the existing service. The below block is a sample of this, based on an example ODS / API Docker environment composition. Make sure you update the mode (`"``SharedInstance"`, `"YearSpecific"`, or `"DistrictSpecific"`) accordingly.

**docker-compose.yml**

```
# ... above are other services
db-admin:
    image: edfialliance/ods-admin-api-db-admin:${ADMIN_API_DB_TAG}
    environment:
      POSTGRES_USER: "${POSTGRES_USER}"
      POSTGRES_PASSWORD: "${POSTGRES_PASSWORD}"
      API_MODE: <YOUR MODE HERE>
    volumes:
      - vol-db-admin:/var/lib/postgresql/data
    restart: always
    container_name: ed-fi-db-admin
# ... below are other services
```

### .env Settings

Add the following to your environment settings file to support Admin API. Note that when running both Admin App and Admin API, some of these settings may overlap. This is expected, and the same values can be used.

**.env for Admin API**

> ```

Reviewer: Ideally, all code fences should have a language. We could try `ini` here. In the docusaurus config file, need to add 'ini' on line 177, because it is not one of the default languages.

# ADMIN_API_HEALTHCHECK_TEST=/bin/true
#  To add a custom health check, consult the documentation at https://docs.docker.com/compose/compose-file/compose-file-v3/#healthcheck

# The following needs to be set to specify the ODS API endpoint for Admin API to internally connect.
# If user chooses direct connection between ODS API and Admin API within docker network, then set the api internal url as follows
API_INTERNAL_URL = http://${ODS_VIRTUAL_NAME}
```

### Nginx / Gateway Configuration

Update your nginx server configuration to include the Admin API in the reverse proxy.

**default.conf.template**

```
# upstream server config...

server {
    #listen / server config...
    #ssl_certificate config...

    # other locations...

    location /${ADMIN_API_VIRTUAL_NAME} {
        client_max_body_size 20M;
        proxy_pass http://${ADMIN_API_VIRTUAL_NAME};
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-Host $host;
        proxy_set_header X-Forwarded-Port 443;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    }
}
```

## 2\. Relaunch the Docker Composition

After updating the files, restart the docker composition.

> ```

Reviewer: Suugest using `powershell` as the language for all shell commands, unless the command is specifically for bash.

docker compose -f ./compose/your-compose-file.yml --env-file ./.env up -d
```

## 3\. Execute First-Time Configuration

Continue on to [First-Time Configuration for Admin 1.x](../admin-api-1x-for-odsapi-5x-6x/first-time-configuration-for-admin-api-1x.md).

> > [!NOTE]
> > The following is the DockerHub repo for **Admin API v1.4.0 Docker Image** for inclusion in Docker compose:
> >
> > * [edfialliance/ods-admin-api:v1.4.0](https://hub.docker.com/layers/edfialliance/ods-admin-api/v1.4/images/sha256-0a52face1b03e94892dc4d82e05f2fae05e635f1c46b2baf081bbcf2e81d76b1?context=explore)
> >
> > * [edfialliance/ods-admin-api-db:v1.4.0](https://hub.docker.com/layers/edfialliance/ods-admin-api-db/v1.4/images/sha256-70375e3564e9d409dfe8c25d27d504f1b3e15f75e454c9da9f5dc40c30c9d4a3?context=explore)

Reviewer: Should change this to a [docusaurus-style admonition](https://docusaurus.io/docs/3.5.2/markdown-features/admonitions). It will not render the same as in GitHub.
```suggestion
::::note

The following is the DockerHub repo for **Admin API v1.4.0 Docker Image** for inclusion in Docker compose:

* [edfialliance/ods-admin-api:v1.4.0](https://hub.docker.com/layers/edfialliance/ods-admin-api/v1.4/images/sha256-0a52face1b03e94892dc4d82e05f2fae05e635f1c46b2baf081bbcf2e81d76b1?context=explore)
* [edfialliance/ods-admin-api-db:v1.4.0](https://hub.docker.com/layers/edfialliance/ods-admin-api-db/v1.4/images/sha256-70375e3564e9d409dfe8c25d27d504f1b3e15f75e454c9da9f5dc40c30c9d4a3?context=explore)

:::
```
